crates/accelerate/src/synthesis/multi_controlled/mcmt.rs

```diff
@@ -0,0 +1,181 @@
+///         aux_2: ──────────┤ X ├──  <--  implements a controlled operation on all qubits
+///                          └───┘         in the "control" register
+fn ccx_chain<'a>(
+    controls: &'a [usize],
+    auxiliaries: &'a [usize],
+) -> impl DoubleEndedIterator<
+    Item = (
+        PackedOperation,
+        SmallVec<[Param; 3]>,
+        Vec<Qubit>,
+        Vec<Clbit>,
+    ),
+> + 'a {
+    let n = controls.len() - 1; // number of chain elements
+    let indices = std::iter::once((controls[0], controls[1], auxiliaries[0]))
+        .chain((0..n - 1).map(|i| (controls[i + 2], auxiliaries[i], auxiliaries[i + 1])));
+    indices.map(|(ctrl1, ctrl2, target)| {
+        (
+            StandardGate::CCXGate.into(),
+            smallvec![],
+            vec![
+                Qubit(ctrl1 as u32),
+                Qubit(ctrl2 as u32),
+                Qubit(target as u32),
+            ],
+            vec![],
+        )
+    })
+}
+
+/// Implement multi-control, multi-target of a single-qubit gate using a V-chain with
+/// (num_ctrl_qubits - 1) auxiliary qubits.
+/// ``controlled_gate`` here must already be the controlled operation, e.g. if we
+/// call MCMT of X, then it must be a CX gate. This is because I currently don't know how to
+/// nicely map the single-qubit gate to it's controlled version.
+///
+/// For example, 4 controls and 2 target qubits for the Hadamard gate, generates
+///
+///     q_0: ──■──────────────────────────────────■──
+///            │                                  │
+///     q_1: ──■──────────────────────────────────■──
+///            │                                  │
+///     q_2: ──┼────■────────────────────────■────┼──
+///            │    │                        │    │
+///     q_3: ──┼────┼────■──────────────■────┼────┼──
+///            │    │    │  ┌───┐       │    │    │
+///     q_4: ──┼────┼────┼──┤ H ├───────┼────┼────┼──
+///            │    │    │  └─┬─┘┌───┐  │    │    │
+///     q_5: ──┼────┼────┼────┼──┤ H ├──┼────┼────┼──
+///          ┌─┴─┐  │    │    │  └─┬─┘  │    │  ┌─┴─┐
+///     q_6: ┤ X ├──■────┼────┼────┼────┼────■──┤ X ├
+///          └───┘┌─┴─┐  │    │    │    │  ┌─┴─┐└───┘
+///     q_7: ─────┤ X ├──■────┼────┼────■──┤ X ├─────
+///               └───┘┌─┴─┐  │    │  ┌─┴─┐└───┘
+///     q_8: ──────────┤ X ├──■────■──┤ X ├──────────
+///                    └───┘          └───┘
+///
+#[pyfunction]
+#[pyo3(signature = (controlled_gate, num_ctrl_qubits, num_target_qubits, control_state=None))]
+pub fn mcmt_v_chain(
+    py: Python,
+    controlled_gate: OperationFromPython,
+    num_ctrl_qubits: usize,
+    num_target_qubits: usize,
+    control_state: Option<usize>,
+) -> PyResult<CircuitData> {
+    let packed_controlled_gate = controlled_gate.operation;
+    let num_qubits = if num_ctrl_qubits > 1 {
+        2 * num_ctrl_qubits - 1 + num_target_qubits
+    } else {
+        1 + num_target_qubits // we can have 1 control and multiple targets
+    };
```

Comment:
So here it seems that 0-controlled MCMT-gates are not allowed. Is this handled somewhere?

Reply:
I added a check 👍🏻